domain/infosync/tiflash_manager.go

```diff
@@ -57,22 +58,113 @@ type TiFlashPlacementManager interface {
 	GetPDRegionRecordStats(ctx context.Context, tableID int64, stats *helper.PDRegionStats) error
 	// GetStoresStat gets the TiKV store information by accessing PD's api.
 	GetStoresStat(ctx context.Context) (*helper.StoresStat, error)
-	// Close is to close TiFlashPlacementManager
+	// GetTiFlashProgress calculates TiFlash replica progress
+	GetTiFlashProgress(tableID int64, replicaCount uint64, TiFlashStores map[int64]helper.StoreStat) (string, error)
+	// UpdateTiFlashProgressCache updates tiflashProgressCache
+	UpdateTiFlashProgressCache(tableID int64, progress string)
+	// GetTiFlashProgressFromCache gets tiflash replica progress from tiflashProgressCache
+	GetTiFlashProgressFromCache(tableID int64) string
+	// DeleteTiFlashProgressFromCache delete tiflash replica progress from tiflashProgressCache
+	DeleteTiFlashProgressFromCache(tableID int64)
+	// CleanTiFlashProgressCache clean progress cache
+	CleanTiFlashProgressCache()
+	// Close is to close TiFlashReplicaManager
 	Close(ctx context.Context)
 }
 
-// TiFlashPDPlacementManager manages placement with pd for TiFlash.
-type TiFlashPDPlacementManager struct {
-	etcdCli *clientv3.Client
+// TiFlashReplicaManagerCtx manages placement with pd and replica progress for TiFlash.
+type TiFlashReplicaManagerCtx struct {
+	etcdCli              *clientv3.Client
+	sync.RWMutex         // protect tiflashProgressCache
+	tiflashProgressCache map[int64]string
 }
 
-// Close is called to close TiFlashPDPlacementManager.
-func (m *TiFlashPDPlacementManager) Close(ctx context.Context) {
+// Close is called to close TiFlashReplicaManagerCtx.
+func (m *TiFlashReplicaManagerCtx) Close(ctx context.Context) {
 
 }
 
+func getTiFlashPeerWithoutLagCount(tiFlashStores map[int64]helper.StoreStat, tableID int64) (int, error) {
+	// storeIDs -> regionID, PD will not create two peer on the same store
+	var flashPeerCount int
+	for _, store := range tiFlashStores {
+		regionReplica := make(map[int64]int)
+		err := helper.CollectTiFlashStatus(store.Store.StatusAddress, tableID, &regionReplica)
+		if err != nil {
+			logutil.BgLogger().Error("Fail to get peer status from TiFlash.",
+				zap.Int64("tableID", tableID))
+			return 0, err
+		}
+		flashPeerCount += len(regionReplica)
+	}
+	return flashPeerCount, nil
+}
+
+// GetTiFlashProgress return truncated string to avoid float64 comparison.
```

Comment:
Not sure why do we need a string for it. If we would like to know whether the progress is 1, comparing with float64 1 should be precise enough?

Reply:
Before this PR, need to compare progress that is not 1. But in this PR, we don't need to compare, so that can use float64 directly.